opencog/persist/sql/test_pg_storage.cc

```diff
@@ -0,0 +1,492 @@
+
+#include <opencog/persist/sql/SQLBackingStore.h>
+#include <opencog/persist/sql/PGAtomStorage.h>
+
+#include "libpq-fe.h"
+
+using namespace opencog;
+
+const int DONT_EXPAND_DBNAME = 0;
+
+
+void print_time()
+{
+  time_t rawtime;
+  struct tm * timeinfo;
+  char buffer[80];
+
+  time (&rawtime);
+  timeinfo = localtime(&rawtime);
+
+  strftime(buffer,80,"%d-%m-%Y %I:%M:%S",timeinfo);
+  fprintf(stdout, "%s\n", buffer);
+}
+
+
+void exit_nicely(PGconn* connection)
+{
+    PQfinish(connection);
+    exit(1);
+}
+
+/*
+CREATE TABLE Atoms (
+	-- The uuid maps to the atom handle.
+	-- Must be unique, and must be non-null. Ergo, primary key.
+	-- Use 64-bit int to match the C++ range.
+	uuid	BIGINT PRIMARY KEY,
+
+	-- The atomspace that this atom belongs to.
+	space BIGINT REFERENCES spaces(space),
+
+	-- Atom type, e.g. Link, Node, etc.
+	type  SMALLINT,
+
+	-- Inlined (simple) truth values
+	tv_type  SMALLINT,
+	stv_mean FLOAT,
+	stv_confidence FLOAT,
+	stv_count DOUBLE PRECISION,
+
+	height SMALLINT,
+
+	-- The node name, non-empty only for nodes
+	name    TEXT,
+
+	-- An array of the outgoing edges; non-empty only for links
+	outgoing BIGINT[],
+
+	-- Force the uniqueness of atoms!!
+	UNIQUE (type, name),
+	UNIQUE (type, outgoing)
+*/
+
+#define INSERT_COUNT 1000
+#define LOOP_COUNT 1000
+#define INSERT_BUFFER_SIZE 2048	
+#define VALUES_BUFFER_SIZE 256	
+#define STATEMENT_BUFFER_SIZE (INSERT_BUFFER_SIZE + (INSERT_COUNT * VALUES_BUFFER_SIZE))
+
+#define MICROSECONDS_PER_SEC ((double) 1000000)
+
+#define USE_TRANSACTIONS true 
+#define DONT_USE_TRANSACTIONS false
+
+#define GROUP_INSERTS true 
+#define DONT_GROUP_INSERTS false
+
+char empty_string[] = "";
+char comma_string[] = ",";
+
+
+
+void insert_atoms_exec(PGconn* connection,
+					   bool use_transactions,
+					   bool group_inserts)
+{
+
+	PGresult *result;
+	char insert_part[INSERT_BUFFER_SIZE];
+	char values_part[VALUES_BUFFER_SIZE];
+	char statement[STATEMENT_BUFFER_SIZE];
+	size_t insert_length;
+	size_t values_length;
+	size_t statement_length;
+
+	snprintf(statement, STATEMENT_BUFFER_SIZE, "DELETE FROM Atoms;");
+	result = PQexec(connection, statement);
+	if (!result || PQresultStatus(result) != PGRES_COMMAND_OK)
+	{
+		fprintf(stderr, "insert_atoms_exec - DELETE FROM Atoms failed\n");
+		fprintf(stderr, "  error = %s\n", PQresultErrorMessage(result));
+		PQclear(result);
+	}
+
+	if (use_transactions)
+		fprintf(stdout,"\ninsert_atoms_exec with Transactions, ");
+	else
+		fprintf(stdout,"\ninsert_atoms_exec NO   Transactions, ");
+	if (group_inserts)
+		fprintf(stdout," insert grouping\n");
+	else
+		fprintf(stdout," NO     grouping\n");
+	fprintf(stdout, "Starting inserts ");
+	print_time();
+	
+	// Get the starting times.
+	std::chrono::steady_clock::time_point begin = std::chrono::steady_clock::now();
+	clock_t start_time = clock() ;
+
+	// Generate the insert part.
+	snprintf(insert_part, INSERT_BUFFER_SIZE, "INSERT INTO Atoms "
```

Comment:
The very newest postgres version 9.5 has a combined INSERT/UPDATE statement:  `UPSERT` -- actually `ON CONFLICT DO UPDATE`  http://www.postgresql.org/docs/9.5/static/sql-insert.html

http://www.craigkerstiens.com/2015/05/08/upsert-lands-in-postgres-9.5/

https://www.compose.io/articles/postgresqls-future-is-looking-upsert/

Using this would be nicer than the current select craziness, but postgres 9,.5 is not available in trusty 14.04 so that makes it kind-of unusable for the short term.